PR: Major overhaul of Readme, contributing and install docs

CAM-Gerlach opened the pull request:
To partially address #6075 as well as a related problem that @csabella and I reproduced that prevented installing from source, as well as more generally make clear our recommendation to use `conda`/Anaconda over `pip` and other distributions and our inability to intervene to solve installation issues with those who choose the latter (as we get many every day on our repo), I've revised the readme, CONTRIBUTING.md and installation documentation accordingly. As part of that, as I'd been hoping to for a while, I took the opportunity to reorganize those sections to eliminate redundancy and outline the alternatives clearly, as well as associated minor cleanup and copyediting. I left the dependancy, etc. and other sections as alone as possible to avoid possible conflicts when forward-porting to `master`.

CAM-Gerlach commented:
@ccordoba12  Any chance we can get this in for 3.2.6? Assuming people actually read it, the updated sections would hopefully help address a significant proportion of the user problems that are reported on our issue tracker, particularly recently. Thanks!

CAM-Gerlach commented:
@jnsebgosselin Your detailed and excellent edits and suggestions were very helpful on the FAQ, so if you get the chance, your input here would be particularly valuable as well. Thanks!

jnsebgosselin commented:
@CAM-Gerlach OK I will do it, probably tonight.

CAM-Gerlach commented:
@jnsebgosselin Thanks, really appreciate it! Once #6137 gets merged it might have conflicts (which is one reason I was hoping to get it merged before), but it shouldn't be all too hard to rebase it if necessary.

jnsebgosselin commented:
[PR#6084_review_JSG.docx](https://github.com/spyder-ide/spyder/files/1608572/PR.6084_review_JSG.docx)

There, I've made my revision (**with track changes on**) in docx this time since GiHub doesn't support uploading odt files. It would not have been practical at all to do it from GitHub directly, so that is why I opted to do this in this format.

Thank you @CAM-Gerlach for all the work you have put into this. I think this is a substantial improvement.

CAM-Gerlach commented:
Thanks! I only have two hours until I need to leave for my conference flight, and I still need to eat, finish packing and get a few things done (never mind sleep...what's that again? xD ) so I'll probably look at it tomorrow—er, no, today...wuts the difference?

> with track changes on

:-)

> since GiHub doesn't support uploading odt files

What? You'd think if any one would support ``odt`` it would be Github, the largest site in the world for open source projects...

jnsebgosselin commented:
@CAM-Gerlach  Maybe wait for @ccordoba12 to go over my revision and comments and to submit his own comments/corrections before changing anything. That will be more efficient and we have the time to do it anyway.

CAM-Gerlach commented:
@jnsebgosselin Sure, good thinking. I just got to my hotel, so I haven't had time to do more than read through it.

ccordoba12 commented on CONTRIBUTING.md:
````diff
@@ -121,12 +130,12 @@ Install our test dependencies:
 ```
 
-or using pip
+or using `pip` (not recommended):
````
I think we should change `not recommended` to `for experts only`

CAM-Gerlach replied:
Done

ccordoba12 commented on README.md:
```diff
@@ -45,10 +45,12 @@ Spyder is a Python development environment with a lot of features:
 * **Find in files**
 
-    Supporting regular expressions and mercurial repositories
+    Supporting regular expressions and Mercurial repositories
```
Please remove `and Mercurial repositories`. That's not true anymore.

CAM-Gerlach replied:
Done

ccordoba12 commented on README.md:
```diff
@@ -69,60 +71,98 @@ This section explains how to install the latest stable release of
+### The Easy/Recommended Way: Anaconda
 
-### On Windows:
+The easiest way to install Spyder for any of our platform, and
```
for any ~~of our~~ platform

CAM-Gerlach replied:
Clarified and fixed to "any of our supported platforms". Obviously, there are plenty of platforms we/Anaconda don't support, but experts could conceivably get it running with one of the advanced methods.

ccordoba12 commented on README.md:
```diff
@@ -69,60 +71,98 @@ This section explains how to install the latest stable release of
+**Windows**
 
-### On GNU/Linux
+Spyder is included in these other scientific Python distributions:
```
Spyder is *also* included

CAM-Gerlach replied:
Done.

ccordoba12 commented on README.md:
```diff
@@ -69,60 +71,98 @@ This section explains how to install the latest stable release of
+desire them, if you encounter installation issues you are unable to
+solve on your own, we recommend you use Anaconda instead, as we are
+generally unable to offer individual assistance for problems specific
+to installing via these alternative distributions.
```
Nicely written!

CAM-Gerlach replied:
Thanks! @jnsebgosselin had some suggestions, which I incorporated along with my revisions, so hopefully you still like it.

ccordoba12 commented on README.md:
```diff
@@ -69,60 +71,98 @@ This section explains how to install the latest stable release of
-- Using the [Anaconda Distribution](http://continuum.io/downloads).
-- Installing from source (see below).
+* [WinPython](https://winpython.github.io/)
+* [Python(x,y)](http://python-xy.github.io)
```
Let's remove `Python(x,y)` because I think it's unmaintained now.

CAM-Gerlach replied:
Done

ccordoba12 commented on README.md:
```diff
@@ -69,60 +71,98 @@ This section explains how to install the latest stable release of
-You can also install Spyder with the `pip` package manager, which comes by
-default with most Python installations. For that you need to use the
-command:
+Spyder can be obtained through through
```
Spyder can *also* be obtained through ~~through~~

CAM-Gerlach replied:
Done

ccordoba12 commented on README.md:
```diff
@@ -69,60 +71,98 @@ This section explains how to install the latest stable release of
-To upgrade Spyder to its latest version, if it was installed before, you need
-to run
+* You can often get Spyder through your distribution's package
+manager (i.e. `apt-get`, `yum`, etc), or install from source
```
... install *it* ...

CAM-Gerlach replied:
Removed whole last phrase as it was redundant.

ccordoba12 commented on README.md:
```diff
@@ -69,60 +71,98 @@ This section explains how to install the latest stable release of
+(see below).
 
-    pip install --upgrade spyder
+### The Expert/Cross-platform Way: Installing from source with pip
```
Two problems here: i) Anaconda is also cross-platform; ii) pip installs binaries these days. So I think this should be:

> The Experts Way: Installing with pip

CAM-Gerlach replied:
Agreed; likely whoever originally wrote it did so long ago, and was thinking along the lines that with it being from source, it could be installed on any platform Python could (wider than just Anaconda). Changed to *"The Expert Way: Installing with pip"* (minor grammar tweak to the above).

ccordoba12 commented on README.md:
```diff
@@ -69,60 +71,98 @@ This section explains how to install the latest stable release of
+**Warning:** While this installation method is a viable option for
+experienced users, installing Spyder (and other SciPy stack packages)
+with `pip` can encounter a number of tricky issues. While you are welcome
+to try this on your own if you're an expert, we are unable to help if you
```
... we *won't be able* to help you if you

CAM-Gerlach replied:
Completely re-written following @jnsebgosselin 's general suggestions

ccordoba12 commented on README.md:
````diff
@@ -69,60 +71,98 @@ This section explains how to install the latest stable release of
+
+You can also install Spyder with the `pip` package manager, which comes by
+default with most Python installations. To do so, you need to use the command:
+```
````
There's no need for triple quotes here.

CAM-Gerlach replied:
Okay

ccordoba12 commented on README.md:
````diff
@@ -69,60 +71,98 @@ This section explains how to install the latest stable release of
+default with most Python installations. To do so, you need to use the command:
+```
+    pip install spyder
+```
````
Remove these quotes.

CAM-Gerlach replied:
Okay

ccordoba12 commented on README.md:
````diff
@@ -69,60 +71,98 @@ This section explains how to install the latest stable release of
+```
+To upgrade Spyder to its latest version, if it was installed before, you need
+to run:
+```
````
Remove these quotes.

CAM-Gerlach replied:
Okay

ccordoba12 commented on README.md:
````diff
@@ -69,60 +71,98 @@ This section explains how to install the latest stable release of
+to run:
+```
+    pip install --upgrade spyder
+```
````
Remove these quotes.

CAM-Gerlach replied:
Okay

ccordoba12 commented on README.md:
```diff
@@ -69,60 +71,98 @@ This section explains how to install the latest stable release of
 
 **Important note**: This does not install the graphical Python libraries (i.e.
 PyQt5 or PyQt4) that Spyder depends on. Those have to be installed separately
 after installing Python.
```
This is only true now for Python 2, so please reword this note.

CAM-Gerlach replied:
Already done

ccordoba12 commented on README.md:
```diff
@@ -69,60 +71,98 @@ This section explains how to install the latest stable release of
-3. `python bootstrap.py`
+## Running from source
 
+The fastest way to run Spyder is to run form source, hosted on the
```
This is confusing, so I think we should change it to

> You can also run Spyder from its source code, which is hosted on the

CAM-Gerlach replied:
Already reworded to that effect, per @jnsebgosselin 's suggestion as well. Unsure what the original author was thinking.

ccordoba12 commented on README.md:
```diff
@@ -69,60 +71,98 @@ This section explains how to install the latest stable release of
 You may want to do this for fixing bugs in Spyder, adding new
 features, learning how Spyder works or just getting a taste of it.
+Make sure to copy the path listed under the "Clone or Download" button there,
+or just use https://github.com/spyder-ide/spyder.git .
```
Let's remove the last two lines here to simplify the instructions below.

CAM-Gerlach replied:
Already done, @jnsebgosselin thought the same

CAM-Gerlach replied:
Agreed, removed completely.

ccordoba12 commented on README.md:
````diff
@@ -69,60 +71,98 @@ This section explains how to install the latest stable release of
+```
+conda install spyder
+conda remove spyder
+git clone PATH_FROM_SPYDER_REPO
````
git clone https://github.com/spyder-ide/spyder.git

CAM-Gerlach replied:
Done

ccordoba12 commented on README.md:
```diff
@@ -69,60 +71,98 @@ This section explains how to install the latest stable release of
+conda install spyder
+conda remove spyder
+git clone PATH_FROM_SPYDER_REPO
+cd DIR_YOU_CLONED_IT_TO
```
cd spyder

CAM-Gerlach replied:
Done

ccordoba12 commented on README.md:
```diff
@@ -69,60 +71,98 @@ This section explains how to install the latest stable release of
+for experts only and is not recommend, so you'll have to solve any
+problems on your own. See the
+[installation instructions](http://pythonhosted.org/spyder/installation.html)
+for more detail.
```
for more *details*.

CAM-Gerlach replied:
Yup, already caught that actually.

ccordoba12 commented on doc/installation.rst:
```diff
@@ -4,71 +4,50 @@ Installation
 
-* Or using this command (in case you *don't* use any of those scientific
-  distributions)::
+This is the easiest way to install Spyder for any of our platform, and
```
... for any platform ...

CAM-Gerlach replied:
Fixed as mentioned above.

ccordoba12 commented on doc/installation.rst:
```diff
@@ -4,71 +4,50 @@ Installation
 
-The easy way
-~~~~~~~~~~~~
+Spyder is included in these other scientific Python distributions:
```
is *also* included

CAM-Gerlach replied:
Done

ccordoba12 commented on doc/installation.rst:
```diff
@@ -4,71 +4,50 @@ Installation
-in the `Anaconda <https://www.anaconda.com/download/>`_ Python distribution,
-which comes with everything you need to get started in an all-in-one package.
+* `WinPython <https://winpython.github.io/>`_
+* `Python(x,y) <https://code.google.com/p/pythonxy>`_
```
Remove Python(x,y)

CAM-Gerlach replied:
Done

ccordoba12 commented on doc/installation.rst:
```diff
@@ -4,71 +4,50 @@ Installation
+* `Python(x,y) <https://code.google.com/p/pythonxy>`_
 
+You can start using it immediately after installing one of them
+(you only need to install one!), just like with Anaconda.
```
These lines are not needed after removing the Python(x,y) link.

CAM-Gerlach replied:
Yes; removed already.

ccordoba12 commented on doc/installation.rst:
```diff
@@ -232,63 +203,117 @@ Optional modules
+
+**Important note**: This does not install the graphical Python libraries (i.e.
+PyQt5 or PyQt4) that Spyder depends on. Those have to be installed separately
+after installing Python. Please see their documentation for more on that.
```
As I said above, this is only true now for Python 2.

CAM-Gerlach replied:
Yes, fixed already as mentioned above.

ccordoba12 commented on doc/installation.rst:
```diff
@@ -77,77 +56,69 @@ __ http://www.macports.org/ports.php?by=name&substr=spyder
-Installing or running directly from source
-------------------------------------------
+
+The Expert Way: Installing or running directly from source
```
I think this should be

> The Experts Way: Installing with pip

CAM-Gerlach replied:
Changed head per my previous comment and subhead per yours here.

ccordoba12 commented on doc/installation.rst:
```diff
@@ -232,63 +203,117 @@ Optional modules
+
+You can update Spyder by:
+
+* Updating Anaconda (recommended), WinPython, Python(x,y), MacPorts, or
```
Remove Python(x,y)

CAM-Gerlach replied:
Yup, missed that one.

ccordoba12 commented:
@CAM-Gerlach, I did a quick review with my main concerns. I don't have more time to do another pass, so please let me know when you think this is ready and I'll merge it.

CAM-Gerlach commented:
@ccordoba12 Thanks! Though, I can now see why @jnsebgosselin said

> It would not have been practical at all to do it from GitHub directly

In any case, since particularly with your updates to parts that weren't touched originally, would you like me to go ahead with my previously mentioned plan to split off parts of #6137 into more sensible places with less chance for merge conflicts with existing PRs? Specifically, include the doc changes from #6137 in this one instead (since otherwise they'll conflict), and more narrowly scope that one to just be the UI part, i.e. the 3 dialogs plus the help menu option (with another, separate PR updating the issue template and auto-submission text, that could potentially be merged immediately if desired)?

ccordoba12 commented:
> Specifically, include the doc changes from #6137 in this one instead (since otherwise they'll conflict), and more narrowly scope that one to just be the UI part

Sounds good to me. Please go ahead with that plan :+1:

jnsebgosselin commented:
Hey, I was looking again at this PR and I was thinking that maybe there is no need to repeat the same installation instructions in the `README` and in the ` doc/installation.rst`. I was looking at both and it is almost the same!

The reason I'm saying this is, I think the README is a good place to showcase and promote Spyder, add important messages, and link to external resources. But the very long and tedious installation section in the README is kind of killing this. You have to scroll so much before the `Backers` and `Sponsors` sections appear! 

Maybe we could replace the installing section in the README by one single paragraph with 2-3 sentences and a link to the documentation and the FAQ. That would be less work to maintain and also, and this would make our README more appealing.

But maybe there is also some good reasons to have the same info repeated in the `README` and in the ` doc/installation.rst` that I haven't thought about.

What do you guys think?
@CAM-Gerlach  @ccordoba12 @goanpeca

CAM-Gerlach commented:
@jnsebgosselin 

> Maybe we could replace the installing section in the README by one single paragraph with 2-3 sentences and a link to the documentation and the FAQ. That would be less work to maintain and also, and this would make our README more appealing.

I agree; IMO that sounds like a great idea. Although not nearly as detailed as the actual installation document, that section still dominates the README and I noticed that it is indeed duplicative to quite an extent, taking considerable time to maintain both places in this very PR (as it was in a different format, rst vs md, and written in a quite different way. In fact, I updated several things other places before in another PR, but missed it in the readme. One could argue CONTRIBUTING.md is a little bit too to some extent, but that's more narrowly focused on installing the test builds.

I assumed there was some good reason to keep it (there certainly might be) and thus I kept it in, but if not and all are in agreement I will slim it down. Present status, btw, is that I've implemented the appropriate changes from #6137 and some other tweaks and am in the middle of applying your suggestions/corrections, but didn't want to push till that was complete.

CAM-Gerlach commented:
@ccordoba12 @jnsebgosselin  Thanks again for your input; implemented as requested, plus added the relevant changes from #6137 condensed the install section in the stream me and made a number of other minor fixes/improvements.

jnsebgosselin commented:
@CAM-Gerlach Any development about the removal of the installation instruction in the readme?

CAM-Gerlach commented:
@jnsebgosselin 

> condensed the install section in the read me 

^

> Any development about the removal of the installation instruction in the readme?

Yes; namely, I did what you requested haha. It now just has three sentences talking about Anaconda, why you should install that method, and that the others are not supported, plus a few one liner bullets with the other options, and a link to the full install instructions; going from 103 lines down to 25. Let me know what you think.

jnsebgosselin commented:
@CAM-Gerlach  I've just taken a look at the Readme and I wanted to tell you that you did a great job there! It looks a lot nicer now.

Sorry for not giving any feedback sooner, I've been a little overwhelmed at work lately...

csabella commented:
Congratulations!  Looks great!

CAM-Gerlach commented:
No worries @jnsebgosselin ! I've been rather busy too with catching up after the conference, the beginning of the semester and dealing with the flood of recent issue reports that I've hardly been able to contribute nearly as much to Spyder as I should be (though the big ticket item remaining for me for 3.2.7, rebasing #6154 , is waiting on #6207 to get merged, there's plenty of smaller things to work on).
